### PR TITLE
implemented some more optoins to cmds

### PR DESCRIPTION
**Pull request by systemstart**
hi.

i added "-" as a valid input to cat and -R and --reference to chmod.

### CODE REVIEW

File: cmds/cat/cat.go [hunk 1]
```diff
@@ -25,12 +25,18 @@ var (
...
 )
 
-func catFile(w io.Writer, file string) error {
-	f, err := os.Open(file)
-	if err != nil {
-		return err
+func catFile(w io.Writer, file string) (err error) {
+	var f *os.File
+
+	if file == "-" {
```
Comment (rminnich) [inline cmds/cat/cat.go]: no no no no no. There is a name for stdin, and it's /dev/fd/0 or /dev/tty or something else depending on your kernel. We're not doing this, sorry :-)

File: cmds/cat/cat_test.go [hunk 2]
```diff
@@ -59,3 +60,39 @@ func TestCat(t *testing.T) {
 		t.Fatalf("Reading files failed: got %v, want %v", b.Bytes(), someData)
 	}
 }
+
+func TestCatDashInput(t *testing.T) {
```
Comment (rminnich) [inline cmds/cat/cat_test.go]: very cool, nice to see a test! sorry it's for a thing I have to say no to :-)

File: cmds/chmod/chmod.go [hunk 3]
```diff
@@ -13,32 +13,89 @@ package main
```
Comment (rminnich) [inline cmds/chmod/chmod.go]: This looks like a good improvement, but could we get the chmod changes in a separate PR?

File: cmds/chmod/chmod_test.go [hunk 4]
```diff
@@ -62,6 +62,122 @@ func TestChmodSimple(t *testing.T) {
 	}
```
Comment (rminnich) [inline cmds/chmod/chmod_test.go]: Thanks for the test!

File: cmds/cat/cat.go (hunk 1 above)
Comment (systemstart) [reply cmds/cat/cat.go]: hm, the command roadmap includes the comment "Still need - for stdin", i felt encouraged ;)

**Comment by systemstart** [comment]
i removed commit about the dash thing from cat.

please fix the roadmap document, which states "Still need - for stdin" for cat.

**Comment by hugelgupf** [comment]
Hey, I'm really sorry, but I won't have time to take a look until tomorrow or Friday; I have prep to do for a presentation.

**Comment by rminnich** [comment]
I'll try to go through it tomorrow, sorry, I'm just quite worn out today.

**Comment by systemstart** [comment]
don't panic ;)

File: cmds/chmod/chmod.go (hunk 3 above)
Comment (rminnich) [reply cmds/chmod/chmod.go]: Got a question here. I know the GNU programmers have lots of longoptions that are identical to a short option, but I'm not sure I see a point to replicating that practice. Part of the reason they did this, I suspect, is that the long option is a kind of documentation and the GNU longopts structure has no place for a doc string, which the Go flag struct does. So a long opt, called recursive, is somewhat helpful given that you have no other help text. 
Given this, I think we should drop the -recursive flag, since -R does the same thing, and -h will tell you about the -R option anyway. 
Sound OK? We've had some discussions about this at Google in previous years and I think that's our general rule.
Thanks. If you really want that -recursive switch, I guess we'll do it, but I wanted to at least ask.

I guess reference is OK, if a bit wordy, because there is no equivalent short option (or is there?).

**Comment by bensallen** [comment]
@rminnich I find long options are preferable to use when scripting, and short options are preferable for interactive use. Using long options in a script makes that script more readable and quicker to understand. As you mention it’s a type of self-documentation, which doesn’t require having a man page or the usage statement handy while reading through a script.

File: cmds/chmod/chmod.go [hunk 5]
```diff
@@ -13,32 +13,89 @@ package main
...
-	for _, name := range flag.Args()[1:] {
-		if err := os.Chmod(name, os.FileMode(octval)); err != nil {
-			log.Printf("%v", err)
-			exitError = true
+	for _, name := range fileList {
+		if recursive {
+			err := filepath.Walk(name, func(path string,
+				info os.FileInfo,
+				err error) error {
+				return os.Chmod(path, mode)
```
Comment (hugelgupf) [inline cmds/chmod/chmod.go]:
```go
if err := os.Chmod(path, mode); !os.IsNotExist(err) {
  return err
}
return nil
```

if the file got deleted in the meantime, it doesn't matter.

**Comment by rminnich** [comment]
so do we want to push this in now?

**Comment by systemstart** [comment]
about long options and such: i would even like the idea of replacing flags with some more sophisticated library to provide the same parameter sets that gnu land supports(or at least get close to it).

i like that idea because it would make existing scripts written for some other environment work out of the box. existing scripts might be tools or build scripts and supporting them would open u-root to more existing stuff.

but then again: why rewrite existing tools with the same command line parameters?

license issues with existing tools? not written in go?

catch 22: because we can ;)

**Comment by bensallen** [comment]
@systemstart Checkout https://github.com/spf13/pflag, minimal change from flag, but better support for GNU like flags. Example use to make a CLI compatible gzip: https://github.com/bensallen/u-root/blob/add_gzip/cmds/gzip/gzip.go. One caveat is that it doesn't appear to let you just to have a short flag and no long flag.

**Comment by rminnich** [comment]
One thing to keep in mind here: it's at the moment easy to bring in other programs since we use and work with the standard flags package. If we move to our own, that breaks the easy import of programs. We need to be careful not to mess that up.
